### DISCUSSION TITLE
8.4.0

### DISCUSSION THREAD

**Discussion by kayhantolga** [betalgo/openai#591] [Announcements]
### 8.4.0
- Added Stream support for submitToolOutputsToRun, createRun, and createThreadAndRun
- 🎉 With this update, we are now in sync with OpenAI's latest API changes. We shouldn't have any missing features as of now. 

## What's Changed
* 584 the assistantmodifyrequest not math the doc of openai by @kayhantolga in https://github.com/betalgo/openai/pull/585
* Stream support added for submitToolOuputsToRun, createRun and createT… by @kayhantolga in https://github.com/betalgo/openai/pull/589
* 8.4.0 by @kayhantolga in https://github.com/betalgo/openai/pull/590

**Full Changelog**: https://github.com/betalgo/openai/compare/v8.3.0...v8.4.0

<hr /><em>This discussion was created from the release <a href='https://github.com/betalgo/openai/releases/tag/v8.4.0'>8.4.0</a>.</em>

**Comment by Nicelittlewood** [comment]
GOOOOD!!!!

**Comment by Nicelittlewood** [comment]
@kayhantolga i really not understand how to retrieve content when stream a run.

**Reply by kayhantolga** [reply]
In which one? There are 5 different Streamable endpoints.

**Comment by Nicelittlewood** [comment]
I am trying with "RunCreateAsStream", but I will also try soon the "RunSubmitToolOutputsAsStream" 'cause i need FileRetrive and FunctionCall tools.

I did several tests and I understood why I could not get the results.

The EndPoint responds with several objects from OpenAi, but this is not what happens with the library.

The "RunResponse" object does not cover all the cases.

For example:

"thread.message.delta" responds, from OpenAi, with the object "Delta", where the partial message of the stream is contained:

```javascript
{
  "data":{
    "id":"msg_MMMMMMMMMMMMMMM"
    "object":"thread.message.delta"
    "delta":{
      "content":[
        0:{
          "index":0
          "type":"text"
          "text":{
           "value":" il"
          }
        }
      ]
    }
  }
  "event":"thread.message.delta"
}
``` 

"Thread.run.step" contains its specific detail fields inside "Step details"

```javascript
{
"data":{
    "id":"step_SSSSSSSSSSSSSSSSSSSSSS"
    "object":"thread.run.step"
    "created_at":1718284038
    "run_id":"run_XXXXXXXXXXXXXXXXXXXXX"
    "assistant_id":"asst_AAAAAAAAAAAAAAAAAAA"
    "thread_id":"thread_TTTTTTTTTTTTTTTTTTTTTT"
    "type":"message_creation"
    "status":"in_progress"
    "cancelled_at":NULL
    "completed_at":NULL
    "expires_at":1718284635
    "failed_at":NULL
    "last_error":NULL
    "step_details":{...}
    "usage":NULL
  }
  "event":"thread.run.step.in_progress"
}
``` 

each event responds with its own different object type.

it is true that there are IDs to call individual objects, but the execution becomes inefficient and terribly slow.

An adapter should be created for each event that returns the right object or alternatively updated "RunResponse" with all the possible fields that could come from the stream array (Delta, StepDetails...).

**Reply by kayhantolga** [reply]
Ha 🤔 that's actually a bug. I thought it would return the same object (as you can guess, run response object) all the time, but actually, there are multiple different objects returning from the stream. Let me fix it 😮‍💨

**Reply by kayhantolga** [reply]
https://github.com/betalgo/openai/issues/594

**Reply by kayhantolga** [reply]
fixed in 8.5.0
